xla/service/cpu/onednn_matmul_rewriter.cc

```diff
@@ -178,41 +151,155 @@ auto GELUActivation(HloInstruction* instr, HloInstruction** src) {
   Shape new_shape = bcast->shape();
+  // Broadcast instruction has "dimension" parameter along which its input's
+  // dim should not change. For exmaple,
+  //      dot = f32[3,4,5,6] dot(...)
+  //      arg = f32[3,6]{1,0} parameter(0)
+  //      broad = f32[3,4,5,6]{3,2,1,0} broadcast(arg), dimensions={0,3}
+  //      add = f32[3,4,5,6]{3,2,1,0} add(dot, arg)
+  // can be replaced with the following
+  //      arg = f32[3,6]{1,0} parameter(0)
+  //      bitcast = f32[3,1,1,6]{3,2,1,0} bitcast(arg)
+  //      fused = f32[3,4,5,6]{3,2,1,0} custom-call((..., bitcast)
   auto kept_dimensions = bcast->dimensions();
   for (int i = 0; i < new_shape.rank(); i++) {
     if (!absl::c_linear_search(kept_dimensions, i)) {
       new_shape.set_dimensions(i, 1);
     }
   }
 
-  // Remove dimensions with value=1 to match dot dimensions
-  bool more_dims_to_delete = true;
-  while (more_dims_to_delete && new_shape.rank() > dot_dims.size()) {
-    for (int64_t i = 0; i < new_shape.rank(); i++) {
-      if (new_shape.dimensions()[i] == 1) {
-        new_shape.DeleteDimension(i);
-        break;
-      }
+  // If rank(new_shape) > rank(dot), extra dimensions with value = 1 can be
+  // deleted from the new_shape.
+  int64_t rank_difference = new_shape.rank() - dot_shape.rank();
+  auto new_dims = new_shape.dimensions();
+  std::vector<int64_t> dims_to_delete;
+  for (int i = 0; i < rank_difference; ++i) {
+    if (new_dims[i] == 1) {
+      dims_to_delete.push_back(i);
     }
-    more_dims_to_delete = false;
   }
+  new_shape = ShapeUtil::DeleteDimensions(dims_to_delete, new_shape);
 
-  // Validate dimensions
-  auto new_shape_dims = new_shape.dimensions();
-  for (int i = 0; i < dot_dims.size() - 2; i++) {
-    if (new_shape_dims[i] != 1 && new_shape_dims[i] != dot_dims[i]) {
-      return absl::CancelledError("bias cannot be fused - dimensions mismatch");
-    }
+  // New shape for bias should satisfy the condition: rank(new_shape) <=
+  // rank(dot).
+  if (new_shape.rank() > dot_shape.rank()) {
+    return absl::CancelledError(
+        "Bias shape could not be adjusted for a fusion.");
   }
+
   return new_shape;
 };
 
+inline bool IsOperandFusible(HloInstruction* operand, HloInstruction* dot) {
+  // Check if one of binary operand shape is compatible with matmul for fusion.
+  // An operand is fusable if
+  //    1. rank(oeprand) <= rank(dot) and
+  //    2. Starting from the last dim in backward direction, the dimension
+  //       size of operand is either 1 or same to dot.
+  auto operand_dims = operand->shape().dimensions();
+  auto dot_dims = dot->shape().dimensions();
+  if (operand_dims.size() <= dot_dims.size()) {
+    int operand_idx = operand_dims.size() - 1;
+    int dot_idx = dot_dims.size() - 1;
+    for (; operand_idx >= 0; --operand_idx, --dot_idx) {
+      if (operand_dims[operand_idx] != 1 &&
+          operand_dims[operand_idx] != dot_dims[dot_idx])
+        return false;
+    }
+    return true;
+  } else {
+    return false;
+  }
```

Comment:
Nit: I think fewer nested code is easier to understand.
```suggestion
  if (operand_dims.size() > dot_dims.size()) return false;
  int operand_idx = operand_dims.size() - 1;
  int dot_idx = dot_dims.size() - 1;
  for (; operand_idx >= 0; --operand_idx, --dot_idx) {
    if (operand_dims[operand_idx] != 1 &&
        operand_dims[operand_idx] != dot_dims[dot_idx])
      return false;
  }
  return true;
```